server/src/main/java/org/opensearch/cluster/service/Throttler.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.service;
+
+import org.opensearch.common.AdjustableSemaphore;
+
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Base class for Throttling logic.
+ * It provides throttling functionality over multiple keys.
+ * It provides the functionality of enable/disable throttling using enableThrottling variable.
+ *
+ * @param <T> the type of key on which we want to do throttling.
+ */
+public class Throttler<T> {
+    protected ConcurrentMap<T, AdjustableSemaphore> semaphores = new ConcurrentHashMap<T, AdjustableSemaphore>();
+
+    private volatile boolean throttlingEnabled;
```

Comment:
this class provides 2 configuration - one is overall throttling enabled / disabled and then per key throttling limit. do we need overall throttling flag ?

Reply:
Earlier idea was to provide one hook for overall throttling enabled and disable. But we are not going to provide that in master throttling. we can remove it from here as well. 

Will remove overall throttling flag.